icu4c/source/common/uloc.cpp

```diff
@@ -848,20 +848,59 @@ uloc_setKeywordValue(const char* keywordName,
                      const char* keywordValue,
                      char* buffer, int32_t bufferCapacity,
                      UErrorCode* status)
+{
+    if (U_FAILURE(*status)) {
+        return -1;
+    }
+
+    if (bufferCapacity <= 1) {
+        *status = U_ILLEGAL_ARGUMENT_ERROR;
+        return 0;
+    }
+
+    int32_t bufLen = (int32_t)uprv_strlen(buffer);
+    if(bufferCapacity<bufLen) {
+        /* The capacity is less than the length?! Is this NUL terminated? */
+        *status = U_ILLEGAL_ARGUMENT_ERROR;
+        return 0;
+    }
+
+    char* keywords = const_cast<char*>(locale_getKeywordsStart(buffer));
+    int32_t baseLen = keywords == nullptr ? bufLen : keywords - buffer;
+    CheckedArrayByteSink sink(keywords == nullptr ? buffer + bufLen : keywords,
+                              bufferCapacity - baseLen - 1);
+    int32_t reslen = ulocimp_setKeywordValue(
+            keywords, keywordName, keywordValue, sink, status);
+
+    if (U_FAILURE(*status)) {
+        return reslen > 0 ? reslen + baseLen : reslen;
+    }
+
+    // See the documentation for this function, it's guaranteed to never
+    // overflow the buffer but instead abort with BUFFER_OVERFLOW_ERROR.
+    U_ASSERT(!sink.Overflowed());
+    U_ASSERT(reslen >= 0);
```

Comment:
Why this assertion? The new function returns 0 if the buffer starts without keywords and keywordValueLen == 0. Is that wrong?

Reply:
Yes, we know that this is what's supposed to happen, this assertion is there to verify that this really is what happens (for `reslen + baseLen` on the following line would be really nasty to debug if there was some bug that caused `reslen` to be negative).